rasa/engine/graph.py

```diff
@@ -0,0 +1,196 @@
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from collections import ChainMap
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List, Optional, Text, Type
+
+from rasa.engine.runner.dask import logger
+import rasa.shared.utils.common
+
+
+@dataclass
+class SchemaNode:
+    """Represents one node in the schema."""
+
+    needs: Dict[Text, Text]
+    uses: Type[GraphComponent]
+    constructor_name: Text
+    fn: Text
+    config: Dict[Text, Any]
+    eager: bool = False
+    is_target: bool = False
+    is_input: bool = False
+    resource_name: Optional[Text] = None  # TODO: type Resource
+
+
+GraphSchema = Dict[Text, SchemaNode]
+
+
+class GraphComponent(ABC):
+    """Interface for any component which will run in a graph."""
+
+    # TODO: This doesn't enforce that it exists in subclasses..
```

Comment:
maybe something like a `classproperty`? Or (as this requires writing a custom decorator) an abstract static method?

Reply:
I couldn't find a way to get the typing to work for an abstract class property. We could do it as a method, but then it just means we have to call it rather than access as attribute.

Reply:
that would be fine to me 👍🏻